### ISSUE TITLE
[Feature Request] Splitting Long Sequences and Potential Modifications to Stateful Encoders

### ISSUE THREAD

**Issue by rloganiv** [allenai/allennlp#2828]
**Is your feature request related to a problem? Please describe.**

Language models are typically trained on copora consisting of long sequences of text. During training, these long sequences typically need to be split into smaller chunks to prevent memory issues (e.g. we need to manually truncate backpropagation), and model state needs to be propagated to subsequent chunks. _Support for splitting long sequences should be added to AllenNLP_. 

Originally, when I spoke to @matt-gardner about this, he recommended to perform this splitting inside a `DatasetReader`. However, to my understanding this approach will cause each chunk to become a seperate ``Instance``, which is problematic since: a. shuffling the data will not preserve the order of the chunks, and b. even if you don't shuffle the data sequences may be split across elements of the batch. _Ideally, data should be shuffleable at the sequence level before it is split into chunks._

As for propagating state, this can currently be achieved using `_EncoderBase` with `stateful=True`, so long as a signal can be emitted to call `reset_states()` at the start of a new sequence. The drawback of this approach is that it requires that all sequences start at the same time, e.g.
```
|-A-|-A-|-A-|-A-|-B-|...
|-C-|pad|pad|pad|-D-|...
                 ^Start of new sequences
```
In this case there will be many chunks which consist mainly of padding, and training will be biased to overfit the long sequence (e.g. sequence A above).

To avoid excess padding, it would be best to start a new sequence as soon as the previous one terminates, e.g.
```
|-A-|-A-|-A-|-A-|-B-|...
|-C-|-D-|...
    ^ D starts as soon as C ends
```
However, there is no appropriate place to call `reset_states()` in this approach. _Ideally, `reset_states()` should be able to reset only some of the internal states for a batch._

**Describe the solution you'd like**

The `reset_states()` issue can be handled by allowing the user to pass an optional `ByteTensor` with dimension `(batch_size,)` which specifies which parts of the internal hidden state to reset, e.g. in the example above, you could pass `[0, 1]` when D begins and `[1, 0]` when B begins. Since the parameter is optional, this would not break backwards compatibility.

As for how/where to split sequences, I don't know that I have a good solution. The way I've handled this issue in my own work is to create a special `SplitIterator` that takes as an additional input a list of fields in the instance to split. You can refer to the implementation here for details: https://gist.github.com/rloganiv/bdfa7c18e004b5cd8c04638a6b4a9c70

My proposed approach has a few issues.
1. It is not very `DataIterator`-esque. Instead of implementing a `_create_batches` method, it overrides `__call__`. Maybe it just needs to be its own kind of `Iterator` and this is okay.
2. It requires specific instructions for how to split each kind of `Field`. This is not easy to maintain, and will not work if a user creates their own custom `Field`. Maybe `Field.__getitem__` methods should be added which allows slicing and returns a `Field` of the same type?

I'd love to hear if anyone has better ideas/alternative approaches for how to handle this issue before I start working on a PR.

**Comment by matt-gardner** [comment]
For reset states, yes, what you describe is the obvious solution, and a PR would be very welcome.

For the splitting issue, I think this is the same as what I talked to @nelson-liu about when he was thinking about language modeling on contiguous text.  The solution we came up with was to have the dataset reader itself handle all of the batching and such, and then yield pre-batched instances, with a special data iterator that just does nothing, passing through the batched text (with a batch size of 1, because the dataset reader already batched things together itself).  The issue is that it's the dataset reader that knows about the full sequence, so you can't re-shuffle the split location, or start another sequence after the current one runs out, anywhere except the dataset reader.  @nelson-liu [started on this](https://github.com/allenai/allennlp/pull/2414), but the PR never got merged, I think because Nelson got distracted with other things.  Can you take a look at that and see what you think?

**Comment by bratao** [comment]
@rloganiv I do have the same problem as you. I'm training an infinite length tagger. 

I see that you was able to elegantly solve this problem and provided the DataIterator. But is the complete solution available anywhere? My main plain with now is how to train and predict using stateful models.
I'm still learning about Pytorch and AllenNLP and this would be a nice head start!

Thank you

**Comment by rloganiv** [comment]
@bratao Happy to help. The code I posted above is (a simplified) version of an iterator I used in this project: https://github.com/rloganiv/kglm-model.

For an example of a model & training config which uses the iterator you can refer to:
- https://github.com/rloganiv/kglm-model/blob/master/kglm/models/kglm_disc.py
- https://github.com/rloganiv/kglm-model/blob/master/experiments/kglm-disc.jsonnet
To handle the issues raised above, I use PyTorch's default RNN layers (lines 88-97) instead of using a child of `_EncoderBase`, and have my model keep track of the state manually (lines 354-360, 416). Also note that the iterator creates the `reset` tensor for you, you don't need to add any fields in your dataset reader.

**Comment by bratao** [comment]
Thank you so much @rloganiv . I will study your code

**Comment by matt-gardner** [comment]
@rloganiv, I'd love to see your work get merged.  What's the next thing to do on this?

**Comment by rloganiv** [comment]
Sorry about the delay. I've been pretty distracted by conferences / final projects the last week. I plan on starting on this next week.

**Comment by rloganiv** [comment]
Finally got around to reading #2414.

> The solution we came up with was to have the dataset reader itself handle all of the batching and such, and then yield pre-batched instances, with a special data iterator that just does nothing, passing through the batched text (with a batch size of 1, because the dataset reader already batched things together itself).

This approach approach makes sense. It will be much easier to maintain than what I propose above, and does not require the `Iterator` create a field (e.g. reset) that is not described in the `DatasetReader`. 

>  The issue is that it's the dataset reader that knows about the full sequence, so you can't re-shuffle the split location, or start another sequence after the current one runs out, anywhere except the dataset reader.

I think this is fine. It might add a little bit of overhead but I am pretty sure all of the shuffling/splitting issues can be addressed by setting `lazy=True` in the dataset reader.

Concrete steps for moving forward:
1. I will submit a PR implementing the reset states solution I described above, then close this issue.
2. I'll work with @nelson-liu to finish up #2414. I've already added some comments to the PR on what I'd like to see changed.
3. (Optional) Since the proposed solution (batching in the dataset reader) is IMO not immediately obvious, I think it would be good to add a simple example to the tutorials. I can write something up for AWD-LSTM.

**Comment by matt-gardner** [comment]
This plan sounds great to me.  For step 3, I'm writing an allennlp course, and it'd be great to have a chapter on how to write code for language modeling with allennlp (it's in my outline).  That's probably the right place for the tutorial you're thinking of.  I can show you the course and what a chapter might look like.  You don't have to commit to writing the whole chapter.
